proofs/goal.ml

```diff
@@ -92,8 +92,8 @@ module V82 = struct
   let same_goal evars1 gl1 evars2 gl2 =
     let evi1 = Evd.find evars1 gl1 in
     let evi2 = Evd.find evars2 gl2 in
-    let c1 = EConstr.Unsafe.to_constr evi1.Evd.evar_concl in
-    let c2 = EConstr.Unsafe.to_constr evi2.Evd.evar_concl in
+    let c1 = EConstr.to_constr ~abort_on_undefined_evars:false evars1 evi1.Evd.evar_concl in
```

Comment:
It really seems like we're going to backtrack on 9f723f14e5342c1303646b5ea7bb5c0012a090ef or what!?

Reply:
Indeed why not to use the native `EConstr` equality [which should by the way perform much better in failure]

Reply:
Oh, we need a new kind of equality that can take 2 evar maps here; don't we?

Reply:
There is already such an equality `Constr.compare_head_gen_with` actually.

Reply:
Indeed I reached to that one, so maybe we should try to fix that properly by adding an extra function to `EConstr` that does proper lazy-wrt-substitution equality?

Reply:
There's also `UnivProblem.eq_constr_univs_infer_with` which is called by `Evarutil.eq_constr_univs_test`, except those look at universes.